File: app/src/main/java/com/amplifyframework/datastore/sample/MainActivity.java
package com.amplifyframework.datastore.sample;

import android.content.Context;
import android.os.Bundle;
import android.view.ViewGroup;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.amplifyframework.rx.RxDataStoreCategoryBehavior;

public class MainActivity extends AppCompatActivity
        implements LocalPresentation.View, RemotePresentation.View {
    private LocalPresentation.Presenter localPresenter;
    private RemotePresentation.Presenter remotePresenter;
    private LogWindow localLogWindow;
    private LogWindow remoteLogWindow;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        bindButtons();
        bindLogWindows();

        Context context = getApplicationContext();

        LocalPresentation.View localView = this;
        RxDataStoreCategoryBehavior dataStore = RxAmplifyFactory.dataStore(context);
        LocalPresentation.PostInteractor localInteractor = new LocalDataStoreInteractor(dataStore);
        this.localPresenter = new LocalPresenter(localInteractor, localView);

        RemotePresentation.View remoteView = this;
        RxAppSyncClient appSync = RxAmplifyFactory.appSync(context);
        RemotePresentation.ApiInteractor remoteInteractor = new RemoteApiInteractor(appSync);
        this.remotePresenter = new RemotePresenter(remoteInteractor, remoteView);
    }

    private void bindButtons() {
        findViewById(R.id.create_local).setOnClickListener(v -> localPresenter.createLocalItems());
        findViewById(R.id.update_local).setOnClickListener(v -> localPresenter.updateLocalItems());
        findViewById(R.id.delete_local).setOnClickListener(v -> localPresenter.deleteLocalItems());
        findViewById(R.id.query_local).setOnClickListener(v -> localPresenter.listLocalItems());
        findViewById(R.id.clear_local).setOnClickListener(v -> localPresenter.clearLocalLog());
        if (!Landscape.isLandscape(this)) {
            findViewById(R.id.stop_everything).setOnClickListener(v -> localPresenter.stopAllLocalActivities());
            findViewById(R.id.begin_subscription).setOnClickListener(v -> localPresenter.startSubscription());
        } else {
            findViewById(R.id.create_remote).setOnClickListener(v -> remotePresenter.createRemotePost());
            findViewById(R.id.update_remote).setOnClickListener(v -> remotePresenter.updateRemotePosts());
            findViewById(R.id.delete_remote).setOnClickListener(v -> remotePresenter.deleteRemotePosts());
            findViewById(R.id.query_remote).setOnClickListener(v -> remotePresenter.listRemotePosts());
            findViewById(R.id.clear_remote).setOnClickListener(v -> remotePresenter.clearRemoteLogs());
        }
    }

    private void bindLogWindows() {
        ViewGroup parent = findViewById(R.id.content);
        this.localLogWindow = new LogWindow(parent, R.id.local_log);
        if (Landscape.isLandscape(this)) {
            this.remoteLogWindow = new LogWindow(parent, R.id.remote_log);
        }
    }

    @Override
    public void onWindowFocusChanged(boolean hasFocus) {
        super.onWindowFocusChanged(hasFocus);
    }

    @Override
    public void displayLocalLogLine(LogLine logLine) {
        runOnUiThread(() -> localLogWindow.add(logLine));
    }

    @Override
    public void clearLocalLineItems() {
        runOnUiThread(() -> localLogWindow.clear());
    }

    @Override
    public void displayRemoteLogLine(LogLine logLine) {
        runOnUiThread(() -> remoteLogWindow.add(logLine));
    }

    @Override
    public void clearRemoteLineItems() {
        runOnUiThread(() -> remoteLogWindow.clear());
    }
}

File: app/src/main/java/com/amplifyframework/datastore/sample/MainActivity.kt
package com.amplifyframework.datastore.sample

import android.os.Bundle
import android.view.View.OnClickListener
import android.view.ViewGroup

import androidx.appcompat.app.AppCompatActivity

import com.amplifyframework.datastore.sample.LocalPresentation.PostInteractor
import com.amplifyframework.datastore.sample.LocalPresentation.Presenter
import com.amplifyframework.datastore.sample.LocalPresentation.View
import com.amplifyframework.datastore.sample.R.id
import com.amplifyframework.datastore.sample.R.layout
import com.amplifyframework.datastore.sample.RemotePresentation.ApiInteractor

class MainActivity : AppCompatActivity(), View, RemotePresentation.View {
    private lateinit var localPresenter: Presenter
    private lateinit var remotePresenter: RemotePresentation.Presenter
    private lateinit var localLogWindow: LogWindow
    private lateinit var remoteLogWindow: LogWindow

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(layout.activity_main)

        bindButtons()
        bindLogWindows()

        val localView: View = this
        val dataStore = RxAmplifyFactory.dataStore(applicationContext)
        val auth = RxAmplifyFactory.auth(applicationContext)
        val localInteractor: PostInteractor = LocalDataStoreInteractor(auth, dataStore)
        localPresenter = LocalPresenter(localInteractor, localView, applicationContext.resources)

        val remoteView: RemotePresentation.View = this
        val appSync = RxAmplifyFactory.appSync(applicationContext)
        val remoteInteractor: ApiInteractor = RemoteApiInteractor(appSync)
        remotePresenter = RemotePresenter(remoteInteractor, remoteView)
    }

    private fun bindButtons() {
        val buttonActions = mutableMapOf(
            Pair(id.create_local, OnClickListener { localPresenter.createLocalItems() }),
            Pair(id.update_local, OnClickListener { localPresenter.updateLocalItems() }),
            Pair(id.delete_local, OnClickListener { localPresenter.deleteLocalItems() }),
            Pair(id.query_local, OnClickListener { localPresenter.listLocalItems() }),
            Pair(id.clear_local, OnClickListener { localPresenter.clearLocalLog() })
        )
        if (!Landscape.isLandscape(this)) {
            buttonActions.putAll(mapOf(
                Pair(id.stop_everything, OnClickListener { localPresenter.stopAllLocalActivities() }),
                Pair(id.begin_subscription, OnClickListener { localPresenter.startSubscription() }),
                Pair(id.sign_in, OnClickListener { localPresenter.signIn() })
            ))
        } else {
            buttonActions.putAll(mapOf(
                Pair(id.create_remote, OnClickListener { remotePresenter.createRemotePost() }),
                Pair(id.update_remote, OnClickListener { remotePresenter.updateRemotePosts() }),
                Pair(id.delete_remote, OnClickListener { remotePresenter.deleteRemotePosts() }),
                Pair(id.query_remote, OnClickListener { remotePresenter.listRemotePosts() }),
                Pair(id.clear_remote, OnClickListener { remotePresenter.clearRemoteLogs() })
            ))
        }
        for ((id, action) in buttonActions) {
            findViewById<android.view.View>(id).setOnClickListener(action)
        }
    }

    private fun bindLogWindows() {
        val parent = findViewById<ViewGroup>(id.content)
        localLogWindow = LogWindow(parent, id.local_log)
        if (Landscape.isLandscape(this)) {
            remoteLogWindow = LogWindow(parent, id.remote_log)
        }
    }

    override fun displayLocalLogLine(logLine: LogLine) = runOnUiThread { localLogWindow.add(logLine) }

    override fun clearLocalLineItems() = runOnUiThread { localLogWindow.clear() }

    override fun displayRemoteLogLine(logLine: LogLine) = runOnUiThread { remoteLogWindow.add(logLine) }

    override fun clearRemoteLineItems() = runOnUiThread { remoteLogWindow.clear() }
}
